nixos/lib/test-driver/test_driver/driver.py

```diff
@@ -101,6 +106,122 @@ def __exit__(self, *_: Any) -> None:
+        ) = new_driver_info.rstrip().split("\n")
+
+        with self.logger.nested("updating machines"):
+            start_cmds = [
+                NixStartScript(start_script) for start_script in start_scripts.split()
+            ]
+            names = [start_cmd.machine_name for start_cmd in start_cmds]
+
+            # delete machines that are no longer part of the test
+            to_del = []
+            for idx, machine in enumerate(self.machines):
+                if machine.name not in names:
+                    if machine.is_up():
+                        self.logger.warning(
+                            f"{machine.name} removed from the test, but it's running so we're not going to shut it down automatically. Call driver.machines[\"{machine.name}\"].stop() and re-run rebuild() to remove."
+                        )
+                    else:
+                        to_del.append(idx)
+            for idx in sorted(to_del, reverse=True):
+                self.logger.info(
+                    f"{self.machines[idx].name} removed from the test. deleting it from the environment"
+                )
+                del globals()[pythonize_name(self.machines[idx].name)]
+                del self.machines[idx]
+
+            # add and change new machines
+            for start_cmd in start_cmds:
+                existing = [
+                    m for m in self.machines if m.name == start_cmd.machine_name
+                ]
+                if len(existing) == 0:
+                    machine = self.create_machine(start_cmd._cmd)
+                    name = pythonize_name(machine.name)
+                    globals()[name] = machine
+                    self.machines.append(machine)
+
+                    self.logger.info(
+                        f'new machine created {start_cmd.machine_name}. start it with driver.machines["{start_cmd.machine_name}"].start()'
+                    )
+                elif len(existing) == 1:
+                    machine = existing[0]
+                    if machine.start_command._cmd == start_cmd._cmd:
+                        self.logger.info(f"{start_cmd.machine_name} unchanged.")
+                    elif not machine.booted:
+                        self.logger.info(f"{start_cmd.machine_name} changed.")
+                        machine.start_command = start_cmd
+                    else:
+                        with self.logger.nested(
+                            f"{start_cmd.machine_name} updated. switching to new configuration"
+                        ):
+                            # TODO: assuming things are at well-known paths isn't great
+                            switch_cmd = (
+                                Path(start_cmd._cmd).parent.parent
+                                / "system"
+                                / "bin"
+                                / "switch-to-configuration"
+                            )
+                            machine.succeed(f"{switch_cmd} test")
```

Comment:
This only works if the host store is forwarded into the VM.
If the VM is completely image based, the new store paths will have to be copied in.
What if we always forward the host store to `/root/host-store` in the guests? No sane program will find it there, so it doesn't affect the purity or performance of the test, but it does allow us to do a `nix copy` just for the purpose of making this work.
This should error out early if https://github.com/NixOS/nixpkgs/blob/883180e6550c1723395a3a342f830bfc5c371f6b/nixos/modules/virtualisation/qemu-vm.nix#L771

Reply:
We don't have access to the nixos configuration here, only the built output. We could check for the presence of the store path, and if it doesn't exist produce a helpful error message that points to this as a likely culprit. Were you imagining something else?

Always forwarding the host store to `/root/host-store` seems like a big change for only this purpose. It seems not very common to set `useNixStoreImage` in tests, and IIUC it's only for performance purposes, so it can be turned off while interactively developing the test.

That being said, my feelings against adding `/root/host-store` are not strong, so if you or other people feel like it is a good change, I'm happy to add it in.